### ISSUE TITLE
Track: Typescript types support

### ISSUE THREAD

**Issue by hugomrdias** [ipfs/js-ipfs#2945]
We will use JSDoc to declare types for the top level API first and internally we will add types incrementally where we feel it adds value.

A TS config will be added to the repo to enable type declaration generation from JSDoc comments. 

Type declaration tests will be added as we see fit and we will add a ts type check job to our CI.

## Repo track list
Done = ✅ 
In Progress = 🚧 
TODO = ⛔ 
| Status | Repo/PR | Owner |
| --- | --- | --- |
|✅  | aegir https://github.com/ipfs/aegir/pull/671 https://github.com/ipfs/aegir/pull/701 | - |
|✅  | ipfs-utils https://github.com/ipfs/js-ipfs-utils/pull/89 | - |
|✅  | interface-datastore https://github.com/ipfs/interface-datastore/pull/56 | - |
|✅  | datastore-core https://github.com/ipfs/js-datastore-core/pull/39 | - |
|✅  | datastore-level https://github.com/ipfs/js-datastore-level/pull/53 | - |
|✅ | datastore-pubsub https://github.com/ipfs/js-datastore-pubsub/pull/60 | @achingbrain |
|✅  | datastore-fs https://github.com/ipfs/js-datastore-fs/pull/62 | - |
|✅ | ipfs-repo https://github.com/ipfs/js-ipfs-repo/pull/275 | @hugomrdias  | 
|✅ | ipfs-unixfs https://github.com/ipfs/js-ipfs-unixfs/pull/114 https://github.com/ipfs/js-ipfs-unixfs/pull/122 | @achingbrain |
|✅ | ipfs-bitswap https://github.com/ipfs/js-ipfs-bitswap/pull/261 | @Gozala |
|✅ | is-ipfs https://github.com/ipfs-shipyard/is-ipfs/pull/39 | @achingbrain |
|✅ | ipns https://github.com/ipfs/js-ipns/pull/106 | @achingbrain |
|✅ | ipfs-block-service https://github.com/ipfs/js-ipfs-block-service/pull/136 | @achingbrain |
|✅  | [cids](https://github.com/multiformats/js-cid) https://github.com/multiformats/js-cid/pull/131 | - |
|✅ | multihash https://github.com/multiformats/js-multihash/pull/104 | - |
|✅  | multibase https://github.com/multiformats/js-multibase/pull/72 | - |
|✅  | [multiaddr](https://github.com/multiformats/js-multiaddr) https://github.com/multiformats/js-multiaddr/pull/159 |@hugomrdias |
|✅ | multicodec https://github.com/multiformats/js-multicodec/pull/70 | - |
|✅  | multihashing-async https://github.com/multiformats/js-multihashing-async/pull/92 | - |
|✅  | libp2p https://github.com/libp2p/js-libp2p/pull/802 | - |
|✅  | libp2p-interfaces https://github.com/libp2p/js-libp2p-interfaces | - |
|⛔ | ipfsd-ctl | |
|✅ | ipfs https://github.com/ipfs/js-ipfs/pull/3550 | @achingbrain |
|✅  | [err-code](https://github.com/IndigoUnited/js-err-code/issues/31) | - |

## Documentation
https://github.com/ipfs/aegir/blob/master/md/ts-jsdoc.md

## Improvement issues
https://github.com/ipfs/aegir/issues/619#issuecomment-732818829

### External issues to track
https://github.com/TypeStrong/typedoc/issues/1248#issuecomment-609114750
https://github.com/microsoft/TypeScript/issues/41672

**Comment by bluelovers** [comment]
where the branch of it? or it not yet started?

**Comment by hugomrdias** [comment]
not started yet, i will start in the next couple of weeks

**Comment by bluelovers** [comment]
https://github.com/ipfs/js-ipfs-utils/pull/26

**Comment by dapplion** [comment]
Is it possible to externally contribute to the Typescript support effort?

**Comment by bluelovers** [comment]
@dapplion https://github.com/bluelovers/ws-ipfs/tree/master/packages/ipfs-types

**Comment by hugomrdias** [comment]
> Is it possible to externally contribute to the Typescript support effort?

@dapplion sure just pick a repo in the list and start adding JSDocs types to the top level api and mention me in the PR

**Comment by dapplion** [comment]
> > Is it possible to externally contribute to the Typescript support effort?
> 
> @dapplion sure just pick a repo in the list and start adding JSDocs types to the top level api and mention me in the PR

Thanks! Is it on the roadmap to migrate to Typescript?

**Comment by hugomrdias** [comment]
no

**Comment by Gozala** [comment]
I've spend some time trying to add jsdoc type annotations to the https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs and encountered issue that I am not sure is possible to entype.

From what I can tell [`IPFS.create`](https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs/docs/MODULE.md#ipfscreateoptions) returns `Promise<IPFS>` node implementing all the [core APIs](https://github.com/ipfs/js-ipfs/tree/master/docs/core-api). But they way it's all put together is extremely dynamic in nature:

https://github.com/ipfs/js-ipfs/blob/4e2aa837ec53c4f97058bba388b7a117b8491c0e/packages/ipfs/src/index.js#L3
https://github.com/ipfs/js-ipfs/blob/4e2aa837ec53c4f97058bba388b7a117b8491c0e/packages/ipfs/src/core/index.js#L44-L48

Where `apiManager` does quite a bit of voodoo with Proxies (aside note Proxies in JS are inherently slow due to dynimac dispatch that engines aren't able to optimize, at least spidermonkey isn't so it's best unless absolutely necessary I'd recommend against them)

https://github.com/ipfs/js-ipfs/blob/4e2aa837ec53c4f97058bba388b7a117b8491c0e/packages/ipfs/src/core/api-manager.js#L7-L12
https://github.com/ipfs/js-ipfs/blob/4e2aa837ec53c4f97058bba388b7a117b8491c0e/packages/ipfs/src/core/api-manager.js#L15-L22

So as far as I can tell `IPFS` instance represents merge of `Record<string, Record<Key, infer T>>`

https://github.com/ipfs/js-ipfs/blob/4e2aa837ec53c4f97058bba388b7a117b8491c0e/packages/ipfs/src/core/index.js#L44-L48

I do not believe that can be entyped because number of types that need to be inferred is not bound. That is to say if e.g. update was taking bound number of APIs it merged together it would have being doable, which is different from arbitrary number of api objects (values of the passed dict) that are merged.

For example this is how TS entypes `Object.assign` (which effectively infers up to 4 args and gives up if you pass more):

```ts
interface ObjectConstructor {
    /**
     * Copy the values of all of the enumerable own properties from one or more source objects to a
     * target object. Returns the target object.
     * @param target The target object to copy to.
     * @param source The source object from which to copy properties.
     */
    assign<T, U>(target: T, source: U): T & U;

    /**
     * Copy the values of all of the enumerable own properties from one or more source objects to a
     * target object. Returns the target object.
     * @param target The target object to copy to.
     * @param source1 The first source object from which to copy properties.
     * @param source2 The second source object from which to copy properties.
     */
    assign<T, U, V>(target: T, source1: U, source2: V): T & U & V;

    /**
     * Copy the values of all of the enumerable own properties from one or more source objects to a
     * target object. Returns the target object.
     * @param target The target object to copy to.
     * @param source1 The first source object from which to copy properties.
     * @param source2 The second source object from which to copy properties.
     * @param source3 The third source object from which to copy properties.
     */
    assign<T, U, V, W>(target: T, source1: U, source2: V, source3: W): T & U & V & W;

    /**
     * Copy the values of all of the enumerable own properties from one or more source objects to a
     * target object. Returns the target object.
     * @param target The target object to copy to.
     * @param sources One or more source objects from which to copy properties
     */
    assign(target: object, ...sources: any[]): any;
}

**Comment by Gozala** [comment]
Turns out I was wrong and there is a way to make the typing work

<img width="1316" alt="image" src="https://user-images.githubusercontent.com/21236/82392058-c20fab00-99f7-11ea-9db0-8635f0cb8ff7.png">

**Comment by dapplion** [comment]
> no

@hugomrdias  Is there a strong reason or ROI justification in the decision to not consider Typescript over just JSDocs types?

**Comment by achingbrain** [comment]
> Proxies in JS are inherently slow due to dynimac dispatch that engines aren't able to optimize, at least spidermonkey isn't so it's best unless absolutely necessary I'd recommend against them

FWIW pulling the proxy voodoo out is on the list: https://github.com/ipfs/js-ipfs/pull/2762

> Is there a strong reason or ROI justification in the decision to not consider Typescript over just JSDocs types?

We want the tooling support & static analysis goodies that TS enables but there's not a lot of appetite for a ground-up rewrite in another language, particularly at the expense of items in the backlog perceived to be of higher value to the project like connectivity improvements, better content resolution, feature parity with go-IPFS, etc.  It also raises the bar to entry for contributions - e.g. every TS developer knows JS, but not every JS developer knows TS.  We don't want to have external definitions as it's too easy for them to become out of date when non-TS developers submit PRs, so this endeavour is to have TS types in JSDoc style comments for our external APIs only.

**Comment by dapplion** [comment]
@achingbrain I agree with your point, thanks for the explanation!

**Comment by rvagg** [comment]
review help over at https://github.com/ipld/js-ipld-block/pull/55 please

**Comment by bluelovers** [comment]
looks like still far away

**Comment by bluelovers** [comment]
changelog say `add typeScript support`, but when install `ipfs` didn't see any `.d.ts`

**Comment by BlackGlory** [comment]
@bluelovers To use JSDoc as TypeScript types, the following options need to be added.

tsconfig.json
```json
{
  "compilerOptions": {
    "allowJs": true
  , "maxNodeModuleJsDepth": 1
  , "esModuleInterop": true
  }
}
```

If we don't aim to generate d.ts file, the above content should be included in README.

**Comment by bluelovers** [comment]
should not ask user use allowJs for make ipfs support typescript

should create .d.ts when publish new version

**Comment by onichandame** [comment]
The IPFS type is still missing. `create()` returns the type `Promise<any>` where the `any` should be an IPFS instance with properties files, pubsub, etc.

**Comment by Xmader** [comment]
waiting https://github.com/ipfs/aegir/pull/568 to be merged

**Comment by Gozala** [comment]
I have pull request #3281 that builds upon @Xmader's work, it aims to generate typedefs from the jsdoc comments on releases so that ts projects could pick all up without doing any configuration.

**Comment by achingbrain** [comment]
There's now an RC available that should have types in the tarball.

Can you please try it out and comment here if you find problems?

```console
$ npm install ipfs@next
```

or

```console
$ npm install ipfs@0.50.3-rc.22+a00bb678
```

**Comment by achingbrain** [comment]
This has shipped with `0.51.0`!

**Comment by 5310** [comment]
Does anyone have an example project that uses `ipfs` or `ipfs-core` with TypeScript that I can look at for the config? I have been trying for a week now and it always just results in several _thousand_ type errors even down to a bare repository with nothing but `ipfs-core` being pulled into it.

If it helps, I keep getting errors like `Module 'multiaddr' does not refer to a type, but is used as a type here. Did you mean 'typeof import('multiaddr')'` and `Cannot use namespace 'Multiaddr' as a type.` and so on and so forth.

**Comment by achingbrain** [comment]
@5310 https://github.com/ipfs/js-ipfs/pull/3359 has a couple of examples, it’ll be merged next week - could you take a look and see if it helps/makes it more clear?

**Comment by achingbrain** [comment]
This has shipped so I'm going to close this issue - please open new issues for any specific problems that arise.

Great work everyone!

**Comment by DeedleFake** [comment]
How do I import the types? I'm trying to pass the instance to something but I'm unsure what type to use. I tried importing `Init` and `Start` from `ipfs-core`, but it says that those aren't exported by it. The examples don't actually type anything local. They just import `ipfs` from a Typescript file.

**Comment by Gozala** [comment]
@DeedleFake right now there’s no simple type you can import, because some of the dynamic implementation details that will be fixed by https://github.com/ipfs/js-ipfs/pull/3365

In the meantime you can use [`ReturnType`](https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypetype) based solution.

**Comment by hugomrdias** [comment]
im reopening this issue because now we have a standard method and not all repos follow it yet.

**Comment by gbaranski** [comment]
> @DeedleFake right now there’s no simple type you can import, because some of the dynamic implementation details that will be fixed by #3365
> 
> In the meantime you can use [`ReturnType`](https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypetype) based solution.

How can I use solution with ReturnType? Is there better way than doing like that?
```ts
const ipfs = await IPFS.create();
type IPFSType = typeof ipfs;
```

**Comment by gbaranski** [comment]
Okay, we can use it like that
```ts
type Await<T> = T extends {
    then(onfulfilled?: (value: infer U) => unknown): unknown;
} ? U : T;

type IPFSType = Await<ReturnType<typeof IPFS.create>>;

const ipfs: IPFSType = await IPFS.create();
```

**Comment by Xmader** [comment]
```ts
type Await<T> = T extends PromiseLike<infer U> ? U : T
```

**Comment by gbaranski** [comment]
Actually having those types out of box, I see that you don't have in plans to migrate to typescript, but would you consider that? That would improve developer experience by a lot.

**Comment by achingbrain** [comment]
>  I see that you don't have in plans to migrate to typescript, but would you consider that?

No, see the final paragraph here: https://github.com/ipfs/js-ipfs/issues/2945#issuecomment-631351122

**Comment by gbaranski** [comment]
> It also raises the bar to entry for contributions - e.g. every TS developer knows JS, but not every JS developer knows TS.
But how many TS developers resigned from contributing because they have to read JS without types. And also how many resigned from using the library because of missing types. I don't think that'll be a big problem for JS developers to understand what's going on in Typescript.

**Comment by Gozala** [comment]
@gbaranski I do not believe discussion about migration to TS is going to be productive, even if doing it had no tradeoffs at all (which it does).

What we have been doing however is gradually integrating [TS via JSDoc syntax](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) and generating typedefs from it. That way we are starting to take some advantage of type checker and TS projects can start getting type inference when working with IPFS.

If you feel passionate about improving experience through TS I would encourage you to help (if you can) with our current entypement efforts, there is still planty of surface to cover and plenty of @ts-ignores to address. I believe that is more effective way to address most pain points in regards to TS+IPFS

**Comment by gbaranski** [comment]
> @gbaranski I do not believe discussion about migration to TS is going to be productive, even if doing it had no tradeoffs at all (which it does).
> 
> What we have been doing however is gradually integrating [TS via JSDoc syntax](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) and generating typedefs from it. That way we are starting to take some advantage of type checker, more and more code has types (admittedly in less convenient form) and TS projects can start getting type inference when working with IPFS.
> 
> If you feel passionate about improving experience through TS I would encourage you to help (if you can) with our ongoing entypement efforts, there is still planty of surface to cover and plenty of @ts-ignores to address. I believe that is more effective way to address most pain points in regards to TS+IPFS

I've added my own types for my libp2p project [right here](https://github.com/gbaranski/cryptogram/blob/master/web/types/libp2p.d.ts). But if You want to generate typings using JSDoc, that might be better approach that mine(rewriting API docs which seem outdated). How I can contribute?

**Comment by Gozala** [comment]
@gbaranski given that you have wrote tsdefs for libp2p this might be a good place to start https://github.com/libp2p/js-libp2p/issues/830 which contains bunch of pointers to bunch of followup work that needs to be done that was chosen to not include with initial push.

If you want to take over this https://github.com/libp2p/js-libp2p/pull/831 that would be a great help as well. As per https://github.com/ipfs/aegir/pull/689#issuecomment-742650729 we would like to replace uses of stock event emitter with a typed one https://www.npmjs.com/package/typed-events.ts

If you do end up pursuing things on libp2p end, I would suggest to coordinate with @vasco-santos who is leading the entyping efforts there to make sure we don't step on each others toes.

On the js-ipfs end there is planty of todo as well, I don't have a list handy here but here are few general groups of work:

1. Address many `@ts-ignores` we had to put in place to make progress. You can pick any of them and address it which will move us closer to taking more advantage of TS.
2. There are plenty of dependencies in js-ipfs that are not yet typed. That leads to bunch of things been `any` that also leaks into generated typedefs. Identifying high impact ones and fixing any one of them would be a great help & probably have largest impact on the usability of js-ipfs from TS.
3. Currently [default config sets `"noImplicitAny": false`](https://github.com/ipfs/aegir/blob/f32a87eb595d625892c30d2460c05a92a6f48220/src/config/tsconfig.aegir.json#L7) which is only way we make progress. Long term we want that to turn to `true`. That is directly relates to No 2, since there are just to many untyped deps. Not sure what the best way to make progress here, but help in this area would also be very welcome.

If you do decide to do some work on js-ipfs end (or it's deps) I would highly encourage to read this https://github.com/ipfs/js-ipfs/issues/3413 as we have decided to take slightly change our approach to how do we type annotate things.

Also don't hesitate to reach out to myself.

**Comment by dapplion** [comment]
Long term, say 1 or 2 years will you re-consider the position of not adopting Typescript?

Major projects in the web3 space actively used by DApp developers are already (or in progress) Typescript
- https://github.com/nomiclabs/hardhat - Typescript
- https://github.com/ethers-io/ethers.js - Typescript
- https://github.com/ethereum/web3.js - 15% migrated to Typescript
- https://github.com/graphprotocol/graph-ts - Typescript (AssemblyScript)

If this trend continues how would it affect IPFS adoption in JS projects if it remains the only one without Typescript (or really good types)?

**Comment by achingbrain** [comment]
> Long term, say 1 or 2 years, will you re-consider the position of not adopting Typescript?

Please see the final paragraph here: https://github.com/ipfs/js-ipfs/issues/2945#issuecomment-631351122

I don't think it needs to be all-or-nothing, since:

> really good types

This is what we're targeting here.  If you'd like to help out with this effort, @Gozala has linked to plenty of small-ish pieces that need work, pitching in would be very much appreciated.

**Comment by dapplion** [comment]
> > Long term, say 1 or 2 years, will you re-consider the position of not adopting Typescript?
> 
> Please see the final paragraph here: [#2945 (comment)](https://github.com/ipfs/js-ipfs/issues/2945#issuecomment-631351122)
> 
> I don't think it needs to be all-or-nothing, since:
> 
> > really good types
> 
> This is what we're targeting here. If you'd like to help out with this effort, @Gozala has linked to plenty of small-ish pieces that need work, pitching in would be very much appreciated.

Is JSDocs able (or fully-featured enough) to provide the same very rich types as "Typescript types"?

I would be very happy to contribute to a Typescript effort, but developing JSDocs is not fun at all for me personally.

However, I agree with prioritizing core functionality development over types.

**Comment by hugomrdias** [comment]
@dapplion yes it is and not all types need to be written in jsdocs and most of them aren't already, please check the description on this issue and the PRs/links referenced there for more info.

**Comment by Gozala** [comment]
For anyone interested in helping out moving this forward, we have decided to slightly change our current "top to bottom" strategy to "bottom upwards" strategy , meaning focusing our efforts on entyping dependencies used by js-ipfs first. This was motivated by ripple effects that added types to dependencies causes.

So if you were thinking of helping out best thing is to pick one of the libraries from the list that @hugomrdias will post here (pardon my gentle nudge 😉).

**Comment by hugomrdias** [comment]
the list has been in the OP for quite some time now

**Comment by achingbrain** [comment]
This was shipped in `ipfs@0.55.x` and `ipfs-core@0.6.x` so I think we can close this now.

If there are further issues with the types, new issues can be opened and dealt with case by case.

Thanks everyone, amazing work!
